Assets/WebGLTemplates/WebVR/webvr.js

```
@@ -33,66 +34,98 @@
       testTimeStart = null;
+      return;
     }
 
-    // Wait for Unity to render frame, then submit to vrDisplay.
-    if (msg.detail === "PostRender") {
-      if (vrDisplay && vrDisplay.isPresenting) {
-        vrDisplay.submitFrame();
+    // Wait for Unity to render the frame; then submit the frame to the VR display.
+    if (msg.detail === 'PostRender') {
+      submitNextFrame = vrDisplay && vrDisplay.isPresenting;
+      if (submitNextFrame) {
+        vrDisplay.requestAnimationFrame(onAnimate);
       }
     }
   }
 
   function onToggleVR() {
-    if (!vrDisplay.isPresenting) {
-      console.log('toggle present')
-      onRequestPresent();
-    } else {
-      console.log('toggle exit present')
+    if (vrDisplay && vrDisplay.isPresenting) {
+      console.log('Toggled to exit VR mode');
       onExitPresent();
+    } else {
+      console.log('Toggled to enter VR mode');
+      onRequestPresent();
     }
   }
 
   function onRequestPresent() {
-    vrDisplay.requestPresent([{ source: canvas }]).then(function() {
-      // starts stereo rendering in Unity.
-      console.log('vr request present success!');
+    if (!vrDisplay) {
+      throw new Error('No VR display available to enter VR mode');
+    }
+    if (!vrDisplay.capabilities || !vrDisplay.capabilities.canPresent) {
+      throw new Error('VR display is not capable of presenting');
+    }
+    return vrDisplay.requestPresent([{source: canvas}]).then(function () {
+      // Start stereo rendering in Unity.
+      console.log('Entered VR mode');
       gameInstance.SendMessage('WebVRCameraSet', 'Begin');
+    }).catch(function (err) {
+      console.error('Unable to enter VR mode:', err);
     });
   }
 
-  function onExitPresent() {
-    vrDisplay.exitPresent().then(function () {
-      console.log('vr exit present success!')
+  function onExitPresent () {
+    if (!vrDisplay && !vrDisplay.isPresenting) {
+      throw new Error('No VR display to exit VR mode');
```

I would use a `console.warn()` here and fail silently but it's your call.

---

for the `VRManager`, I've got in a local branch (initially started for this, and then for #104, and so on), I have turned these functions into methods, and this places nicely with it.

I'm cool with removing that for now. thanks for mentioning the inconsistency; it was bothering me too

---

I saw you replaced a different `throw` with a warning but what about this exception?

---

oh, I missed this one. got distracted with the launch logistics today. will resivit this on Tuesday. (I’m PTO tomorrow, and Monday is a US holiday.)

---

No problem. I'll do.